{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aa0f0d95-5e6b-436f-a971-9486d0841329",
   "metadata": {},
   "source": [
    "# autoMLST Wrapper\n",
    "Summary of [autoMLST Wrapper](https://github.com/KatSteinke/automlst-simplified-wrapper) results from project: `[{{ project().name }}]`\n",
    "\n",
    "## Description\n",
    "This report provides an overview of the result from [autoMLST Wrapper](https://github.com/KatSteinke/automlst-simplified-wrapper), a modified version of [autoMLST](https://bitbucket.org/ziemertlab/automlst) tailored for simplified usability. By integrating a straightforward wrapper script, this fork eliminates the need for additional organism selection steps, streamlining the process for users."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d21448bb-afe9-4fdb-b964-2143a0ccd1e4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from IPython.display import display, Markdown, HTML\n",
    "import json\n",
    "import seaborn as sns\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.filterwarnings('ignore', message='.*method overwritten by.*')\n",
    "\n",
    "from itables import to_html_datatable as DT\n",
    "import itables.options as opt\n",
    "opt.css = \"\"\"\n",
    ".itables table td { font-style: italic; font-size: .8em;}\n",
    ".itables table th { font-style: oblique; font-size: .8em; }\n",
    "\"\"\"\n",
    "opt.classes = [\"display\", \"compact\"]\n",
    "opt.lengthMenu = [5, 10, 20, 50, 100, 200, 500]\n",
    "\n",
    "report_dir = Path(\"../\")\n",
    "\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46dec5a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "report_dir = Path(\"../\")\n",
    "\n",
    "dependency_version = report_dir / \"metadata/dependency_versions.json\"\n",
    "with open(dependency_version, \"r\") as file:\n",
    "    dependency_version = json.load(file)\n",
    "antismash_version = dependency_version[\"antismash\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53fb8b15-379c-4be1-827e-243602f89d95",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(report_dir / \"automlst_wrapper/final.newick\", \"r\") as f:\n",
    "    data = f.readlines()\n",
    "\n",
    "value_to_replace = [i.split(\":\")[0] for i in data[0].replace(\"(\", \"\").split(\",\")]\n",
    "\n",
    "new_dict = {}\n",
    "df = pd.read_csv(\"../automlst_wrapper/df_genomes_tree.csv\")\n",
    "genome_ids = list(df.genome_id)\n",
    "for g in genome_ids:\n",
    "    for v in value_to_replace:\n",
    "        if v.startswith(g.split(\".\")[0]):\n",
    "            new_dict[v] = g\n",
    "            value_to_replace.remove(v)\n",
    "\n",
    "data = data[0]\n",
    "for k in new_dict.keys():\n",
    "    data = data.replace(k, new_dict[k])\n",
    "\n",
    "outfile = Path(\"assets/data/final_corrected.newick\")\n",
    "outfile.parent.mkdir(parents=True, exist_ok=True)\n",
    "with open(outfile, \"w\") as f:\n",
    "    f.write(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b629d79b",
   "metadata": {},
   "source": [
    "## Visualization\n",
    "The tree visualization represents the phylogenetic relationships between various strains of the genus. This visualization aids in understanding the genetic diversity and evolutionary history of these genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6503227-61f9-44e1-8080-6d3c8266e02f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "  library(\"treeio\")\n",
    "  library(\"ggtree\")\n",
    "  library(\"tidyverse\")\n",
    "  library(\"ggstar\")\n",
    "  library(\"ggnewscale\")\n",
    "  library(\"ggtreeExtra\")\n",
    "  library(\"phangorn\")\n",
    "  library(\"RColorBrewer\")\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea15c849-3803-4876-9a91-f38d0e32e337",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R  -w 800 -h 800\n",
    "tree <- read.tree(\"assets/data/final_corrected.newick\")\n",
    "#data <- read.csv(\"../automlst_wrapper/df_genomes_tree.csv\")\n",
    "data <- read.csv(\"../tables/df_gtdb_meta.csv\")\n",
    "\n",
    "# Shorten the Organism column\n",
    "data$Organism_short <- sub(\"^s__([A-Za-z])[a-z]*.*\\\\s\", \"\\\\1. \", data$Organism) # Shorten the genus name\n",
    "data$Organism_short <- sub(\"^s__\", \"\", data$Organism_short)  # Remove 's__'\n",
    "\n",
    "# midpoint root\n",
    "tree <- phangorn::midpoint(tree)\n",
    "tree <- ladderize(reorder(tree))\n",
    "\n",
    "# Add labels to all nodes\n",
    "\n",
    "tree$node.label <- paste0(\"N\", 1:(Nnode(tree) + Ntip(tree)))\n",
    "\n",
    "# Write the tree with internal node IDs to a new Newick file\n",
    "if (!dir.exists(\"assets/iTOL_annotation\")) {\n",
    "  dir.create(\"assets/iTOL_annotation\", recursive = TRUE)\n",
    "}\n",
    "\n",
    "write.tree(tree, file = \"assets/iTOL_annotation/automlst_tree_with_ids.newick\")\n",
    "\n",
    "# Get the unique genera\n",
    "genera <- unique(data$Genus)\n",
    "\n",
    "# Initialize the plot\n",
    "p <- ggtree(tree, \n",
    "            #layout=\"fan\", \n",
    "            size=1, open.angle=5, branch.length='none')\n",
    "\n",
    "# Create a vector of colors\n",
    "colors <- brewer.pal(length(genera), \"Set1\")\n",
    "\n",
    "# Create a mapping from genera to colors\n",
    "genus_to_color <- setNames(colors, genera)\n",
    "\n",
    "# Initialize new columns for color annotation\n",
    "data$tree_color <- NA\n",
    "data$tree_color_label <- NA\n",
    "data$tree_color_MRCA <- NA\n",
    "\n",
    "# Add a clade label for each genus\n",
    "for (genus in genera) {\n",
    "  # Get the tips that belong to this genus\n",
    "  genus_tips <- data$genome_id[data$Genus == genus]\n",
    "  \n",
    "  # Find the MRCA of these tips\n",
    "  mrca_node <- getMRCA(tree, genus_tips)\n",
    "  \n",
    "  # Subtract the number of tips from the MRCA node index\n",
    "  internal_node_index <- mrca_node - Ntip(tree)\n",
    "\n",
    "  # Get the label of the internal node\n",
    "  mrca_label <- tree$node.label[internal_node_index]\n",
    "\n",
    "  # Add the color, label, and MRCA to the new columns\n",
    "  data$tree_color[data$Genus == genus] <- genus_to_color[genus]\n",
    "  data$tree_color_label[data$Genus == genus] <- genus\n",
    "  data$tree_color_MRCA[data$Genus == genus] <- mrca_label\n",
    "\n",
    "  # Highlight this clade\n",
    "  p <- p + geom_hilight(node = mrca_node, fill = genus_to_color[genus], alpha=.6,\n",
    "                        type = \"gradient\", gradient.direction = 'rt',)\n",
    "}\n",
    "\n",
    "# Create a new column that combines the genome_id and Organism fields\n",
    "data$tree_label <- paste(data$genome_id, data$Organism_short, sep=\" - \")\n",
    "\n",
    "# Write the data to a new CSV file\n",
    "write.table(data, file = \"assets/iTOL_annotation/tree_annotation.csv\", sep = \",\", row.names = FALSE)\n",
    "\n",
    "p <- p %<+% data + geom_tippoint(aes(color=Genus), size=3, show.legend = TRUE) + \n",
    "                   geom_tiplab(aes(label=tree_label, offset = 0.5)) + hexpand(.4)\n",
    "\n",
    "# Set the color scale manually\n",
    "p <- p + scale_color_manual(values = genus_to_color)\n",
    "\n",
    "# Move the legend to the bottom\n",
    "p <- p + theme(legend.position = 'bottom')\n",
    "\n",
    "# Combine branch support and label\n",
    "p <- p + geom_text(aes(label=ifelse(isTip, \"\", paste0(label, \" (\", format(round(branch.length, 2), nsmall = 2), \")\"))), vjust=-0.5, hjust=1.1, size=2.8)\n",
    "\n",
    "# Add a scale bar\n",
    "p <- p + geom_treescale(x=0, y=0, offset=0.1)\n",
    "\n",
    "# Display the plot\n",
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eec1beae-aad1-4736-9aec-84fd57eab970",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = Path(\"assets/tables/automlst_tree_table.csv\")\n",
    "outfile.parent.mkdir(parents=True, exist_ok=True)\n",
    "outfile.write_text(df.to_csv(index=False))\n",
    "\n",
    "display(HTML(DT(df.loc[:, [\"genome_id\", \"genus_original\", \"species_original\", \"strain\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"]], scrollX=True)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8fe1ca-30f9-472a-9e01-4e9db61825b8",
   "metadata": {},
   "source": [
    "[Download Table](assets/tables/automlst_tree_table.csv){:target=\"_blank\" .md-button}\n",
    "\n",
    "## Interactive Visualization with iTOL\n",
    "For an enhanced, interactive visualization experience, users are encouraged to download the tree file and the corresponding annotation table. These files can be uploaded to [iTOL (Interactive Tree Of Life)](https://itol.embl.de/), a web-based tool for the display, manipulation, and annotation of phylogenetic trees. Please check the [iTOL help page](https://itol.embl.de/help.cgi) for the upload guide and annotation format.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d09d802",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_annotation = pd.read_csv(\"assets/iTOL_annotation/tree_annotation.csv\")\n",
    "\n",
    "# create label annotation file for iTOL\n",
    "outfile_label = Path(\"assets/iTOL_annotation/iTOL_tree_label.txt\")\n",
    "outfile_label.parent.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "## Write the header to the file\n",
    "with open(outfile_label, 'w') as f:\n",
    "    f.write(\"LABELS\\n\")\n",
    "    f.write(\"SEPARATOR TAB\\n\")\n",
    "    f.write(\"DATA\\n\")\n",
    "\n",
    "## Write the data to the file\n",
    "df_annotation[['genome_id', 'tree_label']].to_csv(outfile_label, sep='\\t', header=False, index=False, mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d7d1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create tree color annotation file for iTOL\n",
    "outfile_color = Path(\"assets/iTOL_annotation/iTOL_tree_color.txt\")\n",
    "\n",
    "with open(outfile_color, 'w') as f:\n",
    "    f.write(\"TREE_COLORS\\n\")\n",
    "    f.write(\"SEPARATOR TAB\\n\")\n",
    "    f.write(\"DATA\\n\")\n",
    "\n",
    "df_annotation[\"tree_color_type\"] = \"range\"\n",
    "color_columns = [\"tree_color_MRCA\", \"tree_color_type\", \"tree_color\", \"tree_color_label\"]\n",
    "df_color = df_annotation[~df_annotation[color_columns].duplicated()][color_columns]\n",
    "#df_color['tree_color_MRCA'] = 'I' + df_color['tree_color_MRCA'].astype(str)\n",
    "\n",
    "## Write the data to the file\n",
    "df_color[color_columns].to_csv(outfile_color, sep='\\t', header=False, index=False, mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f7e34a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "button = f'<a href=\"../assets/iTOL_annotation/automlst_tree_with_ids.newick\" download class=\"md-button\">Download iTOL Tree</a> <a href=\"../{outfile_label}\" download class=\"md-button\">Download iTOL Label</a> <a href=\"../{outfile_color}\" download class=\"md-button\">Download iTOL Color</a>'\n",
    "display(Markdown(button))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3542bf72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_itol_multiple_barchart_annotation(df, dataset_label, output_file, width=50, sep=\"COMMA\"):\n",
    "    # Define the iTOL template\n",
    "    itol_template = \"\"\"DATASET_MULTIBAR\n",
    "#In multi-value bar charts, each ID is associated to multiple numeric values, which are displayed as a stacked or aligned bar chart\n",
    "#lines starting with a hash are comments and ignored during parsing\n",
    "#select the separator which is used to delimit the data below (TAB,SPACE or COMMA).This separator must be used throughout this file (except in the SEPARATOR line, which uses space).\n",
    "\n",
    "#=================================================================\n",
    "#                    MANDATORY SETTINGS                           #\n",
    "#=================================================================#\n",
    "SEPARATOR {sep}\n",
    "DATASET_LABEL,{dataset_label}\n",
    "COLOR,#ff0000\n",
    "#=================================================================#\n",
    "#                    OPTIONAL SETTINGS                            #\n",
    "#=================================================================#\n",
    "WIDTH,1{width}\n",
    "MARGIN,0\n",
    "SHOW_INTERNAL,0\n",
    "HEIGHT_FACTOR,1\n",
    "BAR_SHIFT,0\n",
    "ALIGN_FIELDS,0\n",
    "FIELD_LABELS,{field_labels}\n",
    "FIELD_COLORS,{field_colors}\n",
    "DATASET_SCALE,10,20,30,40,50\n",
    "#=================================================================#\n",
    "#       Actual data follows after the \"DATA\" keyword              #\n",
    "#=================================================================#\n",
    "DATA\n",
    "{data}\n",
    "\"\"\"\n",
    "\n",
    "    # Define the field labels and colors\n",
    "    field_labels = ','.join(df.columns)\n",
    "    \n",
    "    # Get the color palette from seaborn\n",
    "    palette = sns.color_palette(\"Set2\", len(df.columns))\n",
    "\n",
    "    # Convert the RGB values to hexadecimal\n",
    "    field_colors = ','.join([mcolors.rgb2hex(color) for color in palette])\n",
    "\n",
    "    # Format the DataFrame to match the iTOL format\n",
    "    df_itol = df.copy()\n",
    "    df_itol.index.name = 'ID'\n",
    "    df_itol = df_itol.reset_index()\n",
    "    df_itol = df_itol.astype(str)\n",
    "    data = '\\n'.join(df_itol.apply(lambda x: ','.join(x), axis=1))\n",
    "\n",
    "    # Fill the iTOL template with the field labels, colors, and data\n",
    "    itol_annotation = itol_template.format(field_labels=field_labels, field_colors=field_colors, \n",
    "                                           width=width, data=data, sep=sep, dataset_label=dataset_label)\n",
    "\n",
    "    # Write the iTOL annotation to a file\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(itol_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a92f8c3a-69c1-49d1-9307-d97a75de8a8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_itol_heatmap_annotation(df, dataset_label, output_file, strip_width=25, sep=\"COMMA\", color_min=\"#ff0000\", color_max=\"#0000ff\", color_mid=\"#ffff00\"):\n",
    "    # Define the iTOL template\n",
    "    itol_template = \"\"\"DATASET_HEATMAP\n",
    "#In heatmaps, each ID is associated to multiple numeric values, which are displayed as a set of colored boxes defined by a color gradient\n",
    "#lines starting with a hash are comments and ignored during parsing\n",
    "#=================================================================#\n",
    "#                    MANDATORY SETTINGS                           #\n",
    "#=================================================================#\n",
    "#select the separator which is used to delimit the data below (TAB,SPACE or COMMA).This separator must be used throughout this file (except in the SEPARATOR line, which uses space).\n",
    "#SEPARATOR TAB\n",
    "#SEPARATOR SPACE\n",
    "SEPARATOR {sep}\n",
    "\n",
    "#label is used in the legend table (can be changed later)\n",
    "DATASET_LABEL,{dataset_label}\n",
    "\n",
    "#dataset color (can be changed later)\n",
    "COLOR,#ff0000\n",
    "\n",
    "#define labels for each individual field column\n",
    "FIELD_LABELS,{field_labels}\n",
    "\n",
    "#=================================================================#\n",
    "#                    OPTIONAL SETTINGS                            #\n",
    "#=================================================================#\n",
    "\n",
    "\n",
    "#Heatmaps can have an optional Newick formatted tree assigned. Its leaf IDs must exactly match the dataset FIELD_LABELS.\n",
    "#The tree will be used to sort the dataset fields, and will be displayed above the dataset. It can have branch lengths defined.\n",
    "#All newlines and spaces should be stripped from the tree, and COMMA cannot be used as the dataset separator if a FIELD_TREE is provided.\n",
    "#FIELD_TREE (((f1:0.2,f5:0.5):1,(f2_longer_one:0.2,f3:0.3):1.2):0.5,(f4:0.1,f6:0.5):0.8):1.52;\n",
    "#FIELD_TREE (((f1,f5),(f2_longer_one,f3)),(f4,f6));\n",
    "#FIELD_TREE (:0.1,:0.2,(:0.3,:0.4):0.5):0.0;\n",
    "\n",
    "#=================================================================#\n",
    "#     all other optional settings can be set or changed later     #\n",
    "#           in the web interface (under 'Datasets' tab)           #\n",
    "#=================================================================#\n",
    "\n",
    "#left margin, used to increase/decrease the spacing to the next dataset. Can be negative, causing datasets to overlap.\n",
    "MARGIN,0\n",
    "\n",
    "#width of the individual boxes\n",
    "STRIP_WIDTH,{strip_width}\n",
    "\n",
    "#always show internal values; if set, values associated to internal nodes will be displayed even if these nodes are not collapsed. It could cause overlapping in the dataset display.\n",
    "SHOW_INTERNAL,0\n",
    "\n",
    "#if a FIELD_TREE is present, it can be hidden by setting this option to 0\n",
    "SHOW_TREE,1\n",
    "\n",
    "#define the heatmap gradient colors. Values in the dataset will be mapped onto the corresponding color gradient.\n",
    "COLOR_MIN,{color_min}\n",
    "COLOR_MAX,{color_max}\n",
    "\n",
    "#you can specify a gradient with three colors (e.g red to yellow to green) by setting 'USE_MID_COLOR' to 1, and specifying the midpoint color\n",
    "USE_MID_COLOR,1\n",
    "COLOR_MID,{color_mid}\n",
    "\n",
    "#Internal tree nodes can be specified using IDs directly, or using the 'last common ancestor' method described in iTOL help pages\n",
    "#=================================================================#\n",
    "#       Actual data follows after the \"DATA\" keyword              #\n",
    "#=================================================================#\n",
    "DATA\n",
    "{data}\n",
    "\"\"\"\n",
    "\n",
    "    # Define the field labels and colors\n",
    "    field_labels = ','.join(df.columns)\n",
    "    \n",
    "    # Format the DataFrame to match the iTOL format\n",
    "    df_itol = df.copy()\n",
    "    df_itol.index.name = 'ID'\n",
    "    df_itol = df_itol.reset_index()\n",
    "    df_itol = df_itol.astype(str)\n",
    "    data = '\\n'.join(df_itol.apply(lambda x: ','.join(x), axis=1))\n",
    "\n",
    "    # Fill the iTOL template with the field labels, colors, and data\n",
    "    itol_annotation = itol_template.format(field_labels=field_labels, color_min=color_min, color_max=color_max, color_mid=color_mid,\n",
    "                                           strip_width=strip_width, data=data, sep=sep, dataset_label=dataset_label)\n",
    "\n",
    "    # Write the iTOL annotation to a file\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(itol_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94807502-d52c-480e-910e-ed5c69f6e577",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_itol_binary_annotation(df, dataset_label, field_shapes, field_colors, output_file, strip_width=25, sep=\"COMMA\", color=\"#ff0000\", height_factor=0.3, symbol_spacing=0):\n",
    "    # Define the iTOL template\n",
    "    itol_template = \"\"\"DATASET_BINARY\n",
    "#lines starting with a hash are comments and ignored during parsing\n",
    "#select the separator which is used to delimit the data below (TAB,SPACE or COMMA).This separator must be used throught this file (except in the SEPARATOR line, which uses space).\n",
    "\n",
    "#SEPARATOR TAB\n",
    "#SEPARATOR SPACE\n",
    "SEPARATOR {sep}\n",
    "\n",
    "#label is used in the legend table (can be changed later)\n",
    "DATASET_LABEL,{dataset_label}\n",
    "\n",
    "#dataset color (can be changed later)\n",
    "COLOR,{color}\n",
    "\n",
    "#Binary datasets can contain one or more values for each node. Each value will be represented by a symbol (defined in FIELD_SHAPES) with corresponding color and label (from FIELD_COLORS and FIELD_LABELS). Possible values (defined under DATA below) for each node are 1 (filled shapes), 0 (empty shapes) and -1 (completely ommited).\n",
    "\n",
    "#define colors for each individual field column (if not defined all symbols will use the main dataset color, defined in COLOR)\n",
    "#shapes for each field column; possible choices are\n",
    "#1: rectangle \n",
    "#2: circle\n",
    "#3: star\n",
    "#4: right pointing triangle\n",
    "#5: left pointing triangle\n",
    "FIELD_LABELS,{field_labels}\n",
    "FIELD_COLORS,{field_colors}\n",
    "FIELD_SHAPES,{field_shapes}\n",
    "\n",
    "#all other optional settings can be set or changed later in the web interface (under 'Datasets' tab)\n",
    "\n",
    "#show internal values; if set, values associated to internal nodes will be displayed even if these nodes are not collapsed. It could cause overlapping in the dataset display.\n",
    "SHOW_INTERNAL,1\n",
    "\n",
    "#left margin, used to increase/decrease the spacing to the next dataset. Can be negative, causing datasets to overlap.\n",
    "MARGIN,0\n",
    "\n",
    "#symbol height factor; Default symbol height will be slightly less than the available space between leaves, but you can set a multiplication factor here to increase/decrease it (values from 0 to 1 will decrease it, values above 1 will increase it)\n",
    "HEIGHT_FACTOR,{height_factor}\n",
    "\n",
    "#increase/decrease the spacing between individual levels, when there is more than one binary level defined \n",
    "SYMBOL_SPACING,{symbol_spacing}\n",
    "\n",
    "#Internal tree nodes can be specified using IDs directly, or using the 'last common ancestor' method described in iTOL help pages\n",
    "#Actual data follows after the \"DATA\" keyword\n",
    "DATA\n",
    "{data}\n",
    "\"\"\"\n",
    "\n",
    "    # Define the field labels and colors\n",
    "    field_labels = ','.join([f\"GCF_{c}\" for c in df.columns])\n",
    "    \n",
    "    # Format the DataFrame to match the iTOL format\n",
    "    df_itol = df.copy()\n",
    "    df_itol.index.name = 'ID'\n",
    "    df_itol = df_itol.reset_index()\n",
    "    df_itol = df_itol.astype(str)\n",
    "    data = '\\n'.join(df_itol.apply(lambda x: ','.join(x), axis=1))\n",
    "    field_shapes = \",\".join(field_shapes)\n",
    "    field_colors = \",\".join(field_colors)\n",
    "\n",
    "    # Fill the iTOL template with the field labels, colors, and data\n",
    "    itol_annotation = itol_template.format(field_labels=field_labels, color=color, field_shapes=field_shapes,\n",
    "                                           field_colors=field_colors, strip_width=strip_width, data=data, \n",
    "                                           sep=sep, symbol_spacing=symbol_spacing, height_factor=height_factor, \n",
    "                                           dataset_label=dataset_label)\n",
    "\n",
    "    # Write the iTOL annotation to a file\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(itol_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e698b489",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty list to store button items\n",
    "button_items = []\n",
    "\n",
    "# Define the path to the antiSMASH summary file\n",
    "antismash_summary_path = Path(f\"../tables/df_antismash_{antismash_version}_summary.csv\")\n",
    "\n",
    "# Check if the antiSMASH summary file exists\n",
    "if antismash_summary_path.exists():\n",
    "    # Read the antiSMASH summary file into a DataFrame\n",
    "    df_antismash = pd.read_csv(antismash_summary_path).set_index(\"genome_id\")\n",
    "    \n",
    "    # Calculate the number of complete BGCs by subtracting the number of BGCs on the contig edge from the total BGC count\n",
    "    df_antismash[\"complete_bgcs\"] = df_antismash[\"bgcs_count\"] - df_antismash[\"bgcs_on_contig_edge\"]\n",
    "    \n",
    "    # Select the 'complete_bgcs' and 'bgcs_on_contig_edge' columns and convert them to integers\n",
    "    df_antismash_completeness = df_antismash.loc[:, [\"complete_bgcs\", \"bgcs_on_contig_edge\"]].fillna(0).astype(int)\n",
    "    \n",
    "    # Define the output file path\n",
    "    outfile = Path(f\"assets/iTOL_annotation/iTOL_antismash_{antismash_version}_completeness.txt\")\n",
    "    \n",
    "    # Create the output directory if it doesn't exist\n",
    "    outfile.parent.mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # Create the iTOL annotation\n",
    "    create_itol_multiple_barchart_annotation(df_antismash_completeness, \"BGC completeness\", outfile)\n",
    "    \n",
    "    # Create a download button for the iTOL annotation\n",
    "    button_link = f'<a href=\"../{outfile}\" download class=\"md-button\">Download antiSMASH barchart</a>'\n",
    "    \n",
    "    # Add the download button to the list of button items\n",
    "    button_items.append(button_link)\n",
    "\n",
    "    # Define the directory where the BIG-SCAPE results are stored\n",
    "    bigscape_dir = report_dir /f\"bigscape/for_cytoscape_antismash_{antismash_version}/\"\n",
    "\n",
    "    # Check if the BIG-SCAPE directory exists\n",
    "    if bigscape_dir.exists():\n",
    "        # Get the paths to the BIG-SCAPE output files\n",
    "        bgc_table = [i for i in bigscape_dir.glob(\"*_df_clusters_0.30.csv\")][0]\n",
    "        gcf_table = [i for i in bigscape_dir.glob(\"*_df_families_0.30.csv\")][0]\n",
    "        mibig_table = [i for i in bigscape_dir.glob(\"*_df_known_0.30.csv\")][0]\n",
    "        gcf_presence_table = [i for i in bigscape_dir.glob(\"*_df_family_presence_0.30.csv\")][0]\n",
    "        network_table = [i for i in bigscape_dir.glob(\"*_df_network_0.30.csv\")][0]\n",
    "\n",
    "        # Read the BIG-SCAPE output files into DataFrames\n",
    "        df_bgcs = pd.read_csv(bgc_table, index_col=0)\n",
    "        df_gcf_presence = pd.read_csv(gcf_presence_table, index_col=0)\n",
    "        df_gcfs = pd.read_csv(gcf_table, index_col=0)\n",
    "        df_mibig = pd.read_csv(mibig_table, index_col=0)\n",
    "        df_network = pd.read_csv(network_table, index_col=0)\n",
    "\n",
    "        # Initialize a new DataFrame for the genomes\n",
    "        df_genomes = pd.DataFrame()\n",
    "        \n",
    "        # Add a 'Genome ID' column to the genomes DataFrame\n",
    "        df_genomes[\"Genome ID\"] = df_antismash.index\n",
    "        \n",
    "        # Set the 'Genome ID' column as the index of the genomes DataFrame\n",
    "        df_genomes = df_genomes.set_index(\"Genome ID\", drop=False)\n",
    "        \n",
    "        # Add a 'BGCs' column to the genomes DataFrame\n",
    "        df_genomes['BGCs'] = df_antismash.loc[df_genomes.index, 'bgcs_count']\n",
    "\n",
    "        # Get the list of unique BIG-SCAPE classes\n",
    "        bigscape_class_list = df_bgcs.bigscape_class.unique()\n",
    "\n",
    "        # Loop over the index of the genomes DataFrame\n",
    "        for i in df_genomes.index:\n",
    "            # Get the genome ID for the current row\n",
    "            gid = df_genomes.loc[i, 'Genome ID']\n",
    "            \n",
    "            # Check if 'known_family' is in the 'fam_type_0.30' column of the BGCs DataFrame\n",
    "            if \"known_family\" in df_bgcs['fam_type_0.30']:\n",
    "                # Count the number of known families for the current genome and store it in the 'Known BGCs' column\n",
    "                df_genomes.loc[gid, 'Known BGCs'] = df_bgcs[df_bgcs.genome_id == gid].value_counts('fam_type_0.30')['known_family']\n",
    "            \n",
    "            # Check if 'unknown_family' is in the 'fam_type_0.30' column of the BGCs DataFrame\n",
    "            if \"unknown_family\" in df_bgcs['fam_type_0.30']:\n",
    "                # Count the number of unknown families for the current genome and store it in the 'Unknown BGCs' column\n",
    "                df_genomes.loc[gid, 'Unknown BGCs'] = df_bgcs[df_bgcs.genome_id == gid].value_counts('fam_type_0.30')['unknown_family']\n",
    "            \n",
    "            # Count the number of unique BGCs for the current genome and store it in the 'Unique BGCs' column\n",
    "            df_genomes.loc[gid, 'Unique BGCs'] = df_gcf_presence.loc[:, [str(idx) for idx in df_gcfs[df_gcfs.clusters_in_fam==1].index]].sum(1)[gid]\n",
    "            \n",
    "            # Count the number of each BIG-SCAPE class for the current genome\n",
    "            df_bigscape_class_counts = df_bgcs[df_bgcs.genome_id == gid].value_counts('bigscape_class')\n",
    "            \n",
    "            # Loop over the list of unique BIG-SCAPE classes\n",
    "            for bigscape_class in bigscape_class_list:\n",
    "                # Check if the current BIG-SCAPE class is in the index of the BIG-SCAPE class counts DataFrame\n",
    "                if bigscape_class in df_bigscape_class_counts.index:\n",
    "                    # Store the count of the current BIG-SCAPE class in the genomes DataFrame\n",
    "                    df_genomes.loc[gid, bigscape_class] = df_bigscape_class_counts[bigscape_class]\n",
    "        \n",
    "        # create absence presence matrix\n",
    "        df_presence = df_gcf_presence.replace(0, -1)\n",
    "        for gcf_id in df_gcfs.index:\n",
    "            fam_type = df_gcfs.loc[gcf_id, \"fam_type\"]\n",
    "            if fam_type == \"unknown_family\":\n",
    "                df_presence[str(gcf_id)] = df_presence[str(gcf_id)].replace(1, 0)\n",
    "        color = \"#ff0000\"\n",
    "        shape = 1\n",
    "        field_colors = [color for i in df_presence.columns]\n",
    "        field_shapes = [str(shape) for i in df_presence.columns]\n",
    "        \n",
    "        # write absence presence matrix\n",
    "        outfile = Path(f\"assets/iTOL_annotation/iTOL_BiG-SCAPE_presence_antismash_{antismash_version}.txt\")\n",
    "        outfile.parent.mkdir(parents=True, exist_ok=True)\n",
    "        create_itol_binary_annotation(df_presence, \"GCF presence\", field_shapes, field_colors, outfile, sep=\"COMMA\", color=\"#ff0000\")\n",
    "        \n",
    "        # Create a download button for the iTOL annotation\n",
    "        button_link = f'<a href=\"../{outfile}\" download class=\"md-button\">Download BiG-SCAPE GCF presence</a>'\n",
    "        \n",
    "        # Add the download button to the list of button items\n",
    "        button_items.append(button_link)\n",
    "        \n",
    "        # Get the column names of the genomes DataFrame excluding 'Genome ID', 'Unique BGCs', and 'BGCs'\n",
    "        column_names = df_genomes.drop(['Genome ID', 'Unique BGCs', 'BGCs'], axis=1).columns\n",
    "\n",
    "        # Select the columns with the obtained names, fill NA values with 0, and convert the data to integers\n",
    "        df_bigscape_class_summary = df_genomes.loc[:, column_names].fillna(0).astype(int)\n",
    "\n",
    "        # Define the output file path\n",
    "        outfile = Path(f\"assets/iTOL_annotation/iTOL_BiG-SCAPE_class_antismash_{antismash_version}.txt\")\n",
    "\n",
    "        # Create the output directory if it doesn't exist\n",
    "        outfile.parent.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "        # Create the iTOL annotation\n",
    "        create_itol_multiple_barchart_annotation(df_bigscape_class_summary, \"BiG-SCAPE Class\", outfile)\n",
    "\n",
    "        # Create a download button for the iTOL annotation\n",
    "        button_link = f'<a href=\"../{outfile}\" download class=\"md-button\">Download BiG-SCAPE barchart</a>'\n",
    "\n",
    "        # Add the download button to the list of button items\n",
    "        button_items.append(button_link)\n",
    "\n",
    "display(Markdown(\" \".join(button_items)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7672e32",
   "metadata": {},
   "source": [
    "## References\n",
    "<font size=\"2\">\n",
    "\n",
    "- Letunic I and Bork P (2021) Nucleic Acids Res doi: [10.1093/nar/gkab301](https://doi.org/10.1093/nar/gkab301) Interactive Tree Of Life (iTOL) v5: an online tool for phylogenetic tree display and annotation\n",
    "- **G Yu**, DK Smith, H Zhu, Y Guan, TTY Lam<sup>\\*</sup>. ggtree: an\n",
    "    R package for visualization and annotation of phylogenetic trees\n",
    "    with their covariates and other associated data. ***Methods in\n",
    "    Ecology and Evolution***. 2017, 8(1):28-36. doi:\n",
    "    [10.1111/2041-210X.12628](https://doi.org/10.1111/2041-210X.12628)\n",
    "\n",
    "{% for i in project().rule_used['automlst-wrapper']['references'] %}\n",
    "- *{{ i }}*\n",
    "{% endfor %}\n",
    "</font>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
