{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aa0f0d95-5e6b-436f-a971-9486d0841329",
   "metadata": {},
   "source": [
    "# eggNOG-Roary\n",
    "Summary of [eggNOG-Roary](link) results from project: `[{{ project().name }}]` \n",
    "\n",
    "## Description\n",
    "Functional annotation from of roary results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d21448bb-afe9-4fdb-b964-2143a0ccd1e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from IPython.display import display, Markdown, HTML\n",
    "import json\n",
    "import altair as alt\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "from itables import to_html_datatable as DT\n",
    "import itables.options as opt\n",
    "opt.css = \"\"\"\n",
    ".itables table td { font-style: italic; font-size: .8em;}\n",
    ".itables table th { font-style: oblique; font-size: .8em; }\n",
    "\"\"\"\n",
    "opt.classes = [\"display\", \"compact\"]\n",
    "opt.lengthMenu = [5, 10, 20, 50, 100, 200, 500]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53fb8b15-379c-4be1-827e-243602f89d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### UPDATE\n",
    "report_dir = Path(\"../\")\n",
    "project_name = report_dir.resolve().stem\n",
    "roary_interim = Path(f\"../../../interim/roary/{project_name}/\")\n",
    "eggnog_table_path = Path(f\"../../../interim/eggnog_roary/{project_name}/{project_name}.emapper.annotations\")\n",
    "#df_gtdb_gtdbtk = pd.read_csv(\"/datadrive/bgcflow/data/processed/qc_saccharopolyspora/tables/df_gtdb_gtdbtk_meta.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45dd9b44-760a-483e-b215-40d2bab0d0d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gtdb = pd.read_csv(report_dir/ \"tables\" / \"df_gtdb_meta.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39719027-277c-41ff-8cd2-0508c46f1304",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### UPDATE\n",
    "#df_gtdb = df_gtdb_gtdbtk.loc[df_gtdb.index, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97c43bf1-2fab-43a2-a4b1-a5942dde0ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source - https://github.com/SBRG/pymodulon/blob/master/src/pymodulon/gene_util.py\n",
    "def get_cog_dict():\n",
    "    \"\"\"\n",
    "    Get COG dict\n",
    "    \"\"\"\n",
    "\n",
    "    cog_dict = {\n",
    "        \"A\": \"RNA processing and modification\",\n",
    "        \"B\": \"Chromatin structure and dynamics\",\n",
    "        \"C\": \"Energy production and conversion\",\n",
    "        \"D\": \"Cell cycle control, cell division, chromosome partitioning\",\n",
    "        \"E\": \"Amino acid transport and metabolism\",\n",
    "        \"F\": \"Nucleotide transport and metabolism\",\n",
    "        \"G\": \"Carbohydrate transport and metabolism\",\n",
    "        \"H\": \"Coenzyme transport and metabolism\",\n",
    "        \"I\": \"Lipid transport and metabolism\",\n",
    "        \"J\": \"Translation, ribosomal structure and biogenesis\",\n",
    "        \"K\": \"Transcription\",\n",
    "        \"L\": \"Replication, recombination and repair\",\n",
    "        \"M\": \"Cell wall/membrane/envelope biogenesis\",\n",
    "        \"N\": \"Cell motility\",\n",
    "        \"O\": \"Post-translational modification, protein turnover, and chaperones\",\n",
    "        \"P\": \"Inorganic ion transport and metabolism\",\n",
    "        \"Q\": \"Secondary metabolites biosynthesis, transport, and catabolism\",\n",
    "        \"R\": \"General function prediction only\",\n",
    "        \"S\": \"Function unknown\",\n",
    "        \"T\": \"Signal transduction mechanisms\",\n",
    "        \"U\": \"Intracellular trafficking, secretion, and vesicular transport\",\n",
    "        \"V\": \"Defense mechanisms\",\n",
    "        \"W\": \"Extracellular structures\",\n",
    "        \"X\": \"Mobilome: prophages, transposons\",\n",
    "        \"Y\": \"Nuclear structure\",\n",
    "        \"Z\": \"Cytoskeleton\",\n",
    "        \"-\": \"Not found in COG\"\n",
    "    }\n",
    "\n",
    "    return cog_dict\n",
    "\n",
    "cog_dict = get_cog_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c072b8ea-ab03-498d-8009-556115fffd62",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_eggnog = pd.read_csv(eggnog_table_path, sep=\"\\t\", header=4, index_col=\"#query\").iloc[:-3,:]\n",
    "df_eggnog.index.name = \"locus_tag\"\n",
    "df_eggnog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9f0a91c-69cb-4a78-9f1c-14675d33dbee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_pangene_summary = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69542789-ee90-4ebf-9952-f24080941446",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enggnog_stats(df_genomes, eggnog_dir, cog_dict):\n",
    "    '''\n",
    "    Returns dataframe with eggnog statistics\n",
    "    '''\n",
    "    \n",
    "    df_cog = pd.DataFrame(0, index=df_genomes.index, columns=list(cog_dict.keys()))\n",
    "    df_cog_unique = pd.DataFrame(0, index=df_genomes.index, columns=list(cog_dict.keys()))\n",
    "    \n",
    "    df_eggNOG_PA = pd.DataFrame(index=df_genomes.index)\n",
    "    df_eggNOG_PA_fine = pd.DataFrame(index=df_genomes.index)\n",
    "    \n",
    "    df_eggNOG_meta = pd.DataFrame(columns=[\"Preferred_name\", \"Description\", \"COG_category\", \"COG_catergory_name\"])\n",
    "    \n",
    "    \n",
    "    tsv_file = genome_id + \".emapper.annotations\"\n",
    "    eggnog_table_path = eggnog_dir/ genome_id/ tsv_file\n",
    "    if eggnog_table_path.is_file():\n",
    "            df_eggnog = pd.read_csv(eggnog_table_path, sep=\"\\t\", header=4, index_col=\"#query\").iloc[:-3,:]\n",
    "            df_eggnog.index.name = \"locus_tag\"\n",
    "            for locus_tag in df_eggnog.index:\n",
    "                cog_value = df_eggnog.loc[locus_tag, \"COG_category\"]\n",
    "                if len(cog_value) == 1:\n",
    "                    cog_id = cog_value\n",
    "                    df_cog.loc[genome_id, cog_id] = df_cog.loc[genome_id, cog_id] + 1\n",
    "                    df_cog_unique.loc[genome_id, cog_id] = df_cog_unique.loc[genome_id, cog_id] + 1\n",
    "                else:\n",
    "                    for cog_id in cog_value:\n",
    "                        df_cog.loc[genome_id, cog_id] = df_cog.loc[genome_id, cog_id] + 1 \n",
    "                \n",
    "                max_annot_lvl = df_eggnog.loc[locus_tag, \"max_annot_lvl\"]\n",
    "                eggNOG_OGs_list =  df_eggnog.loc[locus_tag, \"eggNOG_OGs\"].split(\",\")\n",
    "                eggNOG_OG_dict = dict()\n",
    "    \n",
    "                for eggNOG_OG in eggNOG_OGs_list:\n",
    "                    annot_lvl = eggNOG_OG.split(\"@\")[1]\n",
    "                    OG_value = eggNOG_OG.split(\"@\")[0]\n",
    "                    eggNOG_OG_dict[annot_lvl] = OG_value\n",
    "                \n",
    "                fine_annot_lvl = eggNOG_OGs_list[-1].split(\"@\")[1]\n",
    "                fine_OG = eggNOG_OGs_list[-1].split(\"@\")[0]\n",
    "                \n",
    "                if fine_OG not in df_eggNOG_meta.index:\n",
    "                    for col_name in df_eggNOG_meta.columns:\n",
    "                        if col_name == \"COG_catergory_name\":\n",
    "                            if len(cog_value) == 1:\n",
    "                                cog_id = cog_value\n",
    "                                df_eggNOG_meta.loc[fine_OG, col_name] = cog_dict[cog_id]\n",
    "                            else:\n",
    "                                cog_name_list = []\n",
    "                                for cog_id in cog_value:\n",
    "                                    cog_name_list.append(cog_dict[cog_id])\n",
    "                                df_eggNOG_meta.loc[fine_OG, col_name] = \";\".join(cog_name_list)\n",
    "                \n",
    "                        else:\n",
    "                            df_eggNOG_meta.loc[fine_OG, col_name] = df_eggnog.loc[locus_tag, col_name]\n",
    "                    \n",
    "\n",
    "                if fine_OG not in df_eggNOG_PA_fine.columns:\n",
    "                    df_eggNOG_PA_fine.loc[:, fine_OG] = 0\n",
    "                    df_eggNOG_PA_fine.loc[genome_id, fine_OG] = 1\n",
    "                else:\n",
    "                    df_eggNOG_PA_fine.loc[genome_id, fine_OG] = df_eggNOG_PA_fine.loc[genome_id, fine_OG] + 1\n",
    "                \n",
    "                if max_annot_lvl in eggNOG_OG_dict.keys():\n",
    "                    selected_OG = eggNOG_OG_dict[max_annot_lvl]\n",
    "                \n",
    "                if selected_OG not in df_eggNOG_PA.columns:\n",
    "                    df_eggNOG_PA.loc[:, selected_OG] = 0\n",
    "                    df_eggNOG_PA.loc[genome_id, selected_OG] = 1\n",
    "                else:\n",
    "                    df_eggNOG_PA.loc[genome_id, selected_OG] = df_eggNOG_PA.loc[genome_id, selected_OG] + 1\n",
    "                \n",
    "    df_cog = df_cog.reindex(columns=df_cog.sum().sort_values(ascending=False).index)\n",
    "    df_cog_unique = df_cog_unique.reindex(columns=df_cog.sum().sort_values(ascending=False).index)\n",
    "    \n",
    "    df_cog_names = df_cog.rename(columns=cog_dict)\n",
    "    df_cog_unique_names = df_cog_unique.rename(columns=cog_dict)\n",
    "    \n",
    "    df_eggNOG_PA = df_eggNOG_PA.reindex(columns=df_eggNOG_PA.sum().sort_values(ascending=False).index)\n",
    "    df_eggNOG_PA_fine = df_eggNOG_PA_fine.reindex(columns=df_eggNOG_PA_fine.sum().sort_values(ascending=False).index)\n",
    "    df_eggNOG_meta = df_eggNOG_meta.reindex(index=df_eggNOG_PA_fine.columns)\n",
    "    \n",
    "    return df_cog, df_cog_names, df_cog_unique, df_cog_unique_names, df_eggNOG_PA, df_eggNOG_PA_fine, df_eggNOG_meta\n",
    "\n",
    "#df_cog, df_cog_names, df_cog_unique, df_cog_unique_names, df_eggNOG_PA, df_eggNOG_PA_fine, df_eggNOG_meta  = get_enggnog_stats(df_gtdb, eggnog_interim, cog_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ac7958-6298-4eba-bd6a-41c6abca4364",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93efb785-7650-4e29-b442-0e527f4e720e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4d8fe1ca-30f9-472a-9e01-4e9db61825b8",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "<font size=\"2\">\n",
    "\n",
    "{% for i in project().rule_used['eggnog-roary']['references'] %}\n",
    "- *{{ i }}*\n",
    "{% endfor %}\n",
    "\n",
    "</font>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
