{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6784920d-a27d-4ace-9a0c-60039490a2bf",
   "metadata": {},
   "source": [
    "# GECCO\n",
    "Summary of BGCs detected by GECCO in each genome for: `[{{ project().name }}]`\n",
    "\n",
    "## Description\n",
    "> GECCO is a fast and scalable method for identifying putative novel Biosynthetic Gene Clusters (BGCs) in genomic and metagenomic data using Conditional Random Fields (CRFs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "690c5018-664b-4bc3-9481-8671834aed55",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from IPython.display import display, Markdown, HTML\n",
    "import json\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "from itables import to_html_datatable as DT\n",
    "import itables.options as opt\n",
    "opt.css = \"\"\"\n",
    ".itables table td { font-style: italic; font-size: .8em;}\n",
    ".itables table th { font-style: oblique; font-size: .8em; }\n",
    "\"\"\"\n",
    "opt.classes = [\"display\", \"compact\"]\n",
    "opt.lengthMenu = [5, 10, 20, 50, 100, 200, 500]\n",
    "\n",
    "import seaborn as sns\n",
    "cm = sns.light_palette(\"green\", as_cmap=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c996f043-4651-4b0f-9625-b410ae3323af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overlap(query, lookup):\n",
    "    overlap_key = None\n",
    "    overlap_length = 0\n",
    "    query_length = query['end'] - query['start']\n",
    "    target_length = 0\n",
    "    overlap_proportion = 0\n",
    "    is_overlapping = \"False\"\n",
    "\n",
    "    for key, value in lookup.items():\n",
    "        if value['start_pos'] <= query['end'] and query['start'] <= value['end_pos']:\n",
    "            overlap_key = key\n",
    "            overlap_length = min(query['end'], value['end_pos']) - max(query['start'], value['start_pos'])\n",
    "            target_length = value['end_pos'] - value['start_pos']\n",
    "            overlap_proportion = overlap_length / target_length\n",
    "            is_overlapping = [\"True\" for i in overlap_key if i != None][0]\n",
    "\n",
    "    return {\"antismash_target\": overlap_key, \"overlap_length\": overlap_length, \"overlap_proportion\" : overlap_proportion, \"query_length\": query_length, \"target_length\": target_length, \"is_overlapping\" : is_overlapping}\n",
    "\n",
    "def bgc_charts(source, var_selection=\"type\", var_y_axis=\"probability_score\", var_x_axis=\"length\",\n",
    "                  var_tooltip=['cluster_id', 'sequence_id', 'start', 'end', 'average_p', 'max_p',\n",
    "                               'type', 'alkaloid_probability', 'nrp_probability',\n",
    "                               'polyketide_probability', 'ripp_probability', 'saccharide_probability',\n",
    "                               'terpene_probability', 'proteins', 'domains', 'genome_id', 'length',\n",
    "                               'probability_score'],\n",
    "                  var_scatter_x_axis_title=\"BGC Region Length (bp)\",\n",
    "                  var_scatter_y_axis_title=\"GECCO BGC Type Probability (%)\",\n",
    "                  var_bar_x_title=\"BGC Count\",\n",
    "                  var_legend_title=\"GECCO BGC Type\",\n",
    "                  var_dropdown_title=\"GECCO BGC Type\"):\n",
    "    options = [i for i in source[var_selection].unique()]\n",
    "\n",
    "    resize = alt.selection_interval(bind='scales')\n",
    "\n",
    "    base = alt.Chart(source)\n",
    "\n",
    "    labels = [option + ' ' for option in options]\n",
    "\n",
    "    input_dropdown = alt.binding_select(options=options + [None],\n",
    "                                        labels=labels + ['All '],\n",
    "                                        name=f'{var_dropdown_title} ')\n",
    "\n",
    "    selection = alt.selection_point(fields=[var_selection], \n",
    "                                    bind=input_dropdown)\n",
    "\n",
    "    color = alt.condition(\n",
    "        selection,\n",
    "        alt.Color(f'{var_selection}:N').legend(None),\n",
    "        alt.value('lightgray')\n",
    "    )\n",
    "\n",
    "    scatter = base.mark_circle(size=75).encode(\n",
    "        x=alt.X(f'{var_x_axis}:Q', title=var_scatter_x_axis_title),\n",
    "        y=alt.Y(f'{var_y_axis}:Q', title=var_scatter_y_axis_title).axis(format='%'),\n",
    "        color=color,\n",
    "        tooltip=var_tooltip\n",
    "    ).add_params(\n",
    "        selection,\n",
    "    ).properties(\n",
    "        height=400,\n",
    "        width=600\n",
    "    ).add_params(\n",
    "        resize\n",
    "    )\n",
    "\n",
    "    legend = base.mark_circle(size=75).encode(\n",
    "        alt.Y(f'{var_selection}:N', title=var_legend_title).axis(orient='right'),\n",
    "        color=color\n",
    "    )\n",
    "\n",
    "    chart2 = base.mark_bar().encode(\n",
    "        x=alt.X('count()', title=var_bar_x_title),\n",
    "        y=alt.Y(f'{var_y_axis}:Q', title=\"\").axis(format='%').bin(maxbins=30),\n",
    "        color=color\n",
    "    ).add_params(\n",
    "        selection,\n",
    "    ).properties(\n",
    "        height=400,\n",
    "        width=100\n",
    "    ).add_params(\n",
    "        resize\n",
    "    )\n",
    "\n",
    "    return scatter | chart2 | legend"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7369351-a742-4289-b04c-53ea4deb3862",
   "metadata": {},
   "outputs": [],
   "source": [
    "report_dir = Path(\"../\")\n",
    "dependency_version = report_dir / \"metadata/dependency_versions.json\"\n",
    "with open(dependency_version, \"r\") as file:\n",
    "    dependency_version = json.load(file)\n",
    "antismash_version = dependency_version[\"antismash\"]\n",
    "gecco_version = dependency_version[\"gecco\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "781915aa-7135-49ef-a062-58da0343b14e",
   "metadata": {},
   "source": [
    "## GECCO Summary Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eef19d7-abc3-4a61-ae0a-195bf7ce171b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gecco_table = report_dir / f\"gecco/{gecco_version}/gecco_clusters.csv\"\n",
    "source = pd.read_csv(gecco_table)\n",
    "\n",
    "source[\"length\"] = source[\"end\"] - source[\"start\"]\n",
    "\n",
    "probability_category = [c for c in source.columns if \"probability\" in c]\n",
    "for i in source.index:\n",
    "    bgc_type = source.loc[i, \"type\"]\n",
    "    probability_score = {}\n",
    "    for t in bgc_type.split(\";\"):\n",
    "        for category in [c for c in probability_category if t.lower() in c]:\n",
    "            value = source.loc[i, category]\n",
    "            probability_score[category] = value\n",
    "        if t == \"Unknown\":\n",
    "            if len(source.loc[i, probability_category]) > 0:\n",
    "                max_column = source.loc[i, probability_category].idxmax()\n",
    "                max_value = source.loc[i, max_column]\n",
    "                probability_score[max_column] = max_value                \n",
    "    if len(probability_score) > 1:\n",
    "        # Get the key with the max value\n",
    "        max_key = max(probability_score, key=probability_score.get)\n",
    "\n",
    "        # Get the max value\n",
    "        max_value = probability_score[max_key]\n",
    "    elif probability_score == {}:\n",
    "        max_value = np.NaN\n",
    "    else:\n",
    "        max_value = [i for i in probability_score.values()][0]\n",
    "    source.loc[i, \"probability_score\"] = max_value\n",
    "\n",
    "for c in [\"alkaloid_probability\", \"nrp_probability\", \"polyketide_probability\", \"ripp_probability\", \"saccharide_probability\", \"terpene_probability\"]:\n",
    "    if c not in source.columns:\n",
    "        source[c] = np.NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5778ef0a-fa40-4a28-a411-31ebae3e9241",
   "metadata": {},
   "outputs": [],
   "source": [
    "chart1 = bgc_charts(source)\n",
    "chart1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d53c0b9f-bf60-4c9a-8f95-eb4a8539f9d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(HTML(DT(source.loc[:, [\"cluster_id\", \"genome_id\", \"average_p\", \"max_p\", \"type\",\n",
    "                                \"length\", \"probability_score\"]], columnDefs=[{\"className\": \"dt-center\", \"targets\": \"_all\"}], scrollX=True)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "498be89d-f5b1-4554-bda2-762cff850cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "antismash_table = Path(report_dir / f\"tables/df_regions_antismash_{antismash_version}.csv\")\n",
    "\n",
    "if antismash_table.is_file():\n",
    "    display(Markdown(\"## Overlap with antiSMASH BGCs\"))\n",
    "    df_antismash = pd.read_csv(antismash_table).set_index(\"bgc_id\")\n",
    "\n",
    "    data = []\n",
    "    for i in source.index:\n",
    "        accession_id = source.loc[i, \"sequence_id\"]\n",
    "        cluster_id = source.loc[i, \"cluster_id\"]\n",
    "        subset_antismash = df_antismash[df_antismash[\"accession\"] == accession_id]\n",
    "        query = source.loc[i, [\"start\", \"end\"]].to_dict()\n",
    "        lookup = subset_antismash.loc[:, [\"start_pos\", \"end_pos\"]].T.to_dict()\n",
    "        overlap = find_overlap(query, lookup)\n",
    "        overlap[\"cluster_id\"] = cluster_id\n",
    "        data.append(overlap)\n",
    "\n",
    "    antismash_overlap = pd.DataFrame.from_dict(data)\n",
    "    source2 = source.merge(antismash_overlap, left_on=\"cluster_id\", right_on=\"cluster_id\")\n",
    "    source2 = source2[~source2.duplicated()].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86a530a7-9bef-4e43-9e35-317b144a1653",
   "metadata": {},
   "outputs": [],
   "source": [
    "chart2 = \"\"\n",
    "if antismash_table.is_file():\n",
    "    chart2 = bgc_charts(source2, \n",
    "                        var_selection=\"is_overlapping\", \n",
    "                        var_y_axis=\"probability_score\", \n",
    "                        var_x_axis=\"length\",\n",
    "                        var_tooltip=['cluster_id', 'antismash_target', 'overlap_length', 'overlap_proportion', \n",
    "                                     'query_length', 'target_length', 'sequence_id', 'start', 'end', 'average_p', \n",
    "                                     'max_p', 'type', 'alkaloid_probability', 'nrp_probability',\n",
    "                                   'polyketide_probability', 'ripp_probability', 'saccharide_probability',\n",
    "                                   'terpene_probability', 'proteins', 'domains', 'genome_id', 'length',\n",
    "                                   'probability_score', ],\n",
    "                      var_scatter_x_axis_title=\"BGC Region Length (bp)\",\n",
    "                      var_scatter_y_axis_title=\"GECCO BGC Type Probability (%)\",\n",
    "                      var_bar_x_title=\"BGC Count\",\n",
    "                      var_legend_title=\"Overlap\",\n",
    "                      var_dropdown_title=\"Overlap with antiSMASH BGCs?\")\n",
    "chart2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e593bf-97a7-4a9b-bda2-c9aabd25947b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if antismash_table.is_file():\n",
    "    display(HTML(DT(source2.loc[:, [\"cluster_id\", \"genome_id\", \"average_p\", \"max_p\", \"type\",\n",
    "                                \"length\", \"probability_score\", \"antismash_target\", \"overlap_length\", \n",
    "                                \"overlap_proportion\", \"is_overlapping\"]], columnDefs=[{\"className\": \"dt-center\", \"targets\": \"_all\"}], scrollX=True)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d6ef83a-9070-49f0-8430-2d9709b8e814",
   "metadata": {},
   "source": [
    "## References\n",
    "<font size=\"2\">\n",
    "{% for i in project().rule_used['gecco']['references'] %}\n",
    "- {{ i }} \n",
    "{% endfor %}\n",
    "</font>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
