{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aa0f0d95-5e6b-436f-a971-9486d0841329",
   "metadata": {},
   "source": [
    "# BiG-FAM Query\n",
    "Summary of [BiG-FAM Query](https://bigfam.bioinformatics.nl/home) results from project: `[{{ project().name }}]`\n",
    "\n",
    "## Description\n",
    "Query against 1.2 Million BGCs in the BiG-FAM database using BiG-SLICE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d21448bb-afe9-4fdb-b964-2143a0ccd1e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import sqlite3\n",
    "from IPython.display import display, Markdown, HTML\n",
    "\n",
    "import json, yaml\n",
    "import altair as alt\n",
    "import ast\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import networkx as nx\n",
    "import plotly.graph_objects as go\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dd35e5b-bca8-4c8a-9187-ddfd61e1d957",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_bigfam_network(df, rank=0):\n",
    "    \"\"\"\n",
    "    Build a networkx graph from bigscape df network\n",
    "    \"\"\"\n",
    "    df[df['rank'] == rank]\n",
    "    G = nx.from_pandas_edgelist(df, source='bgc_id', target='gcf_id', edge_attr=['membership_value',\n",
    "           'rank'], edge_key='bigfam_edge_id')\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1a957a4-f659-436d-9497-bf5933572124",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_bigfam_models(G, df, columns = ['bgc_member', 'chemical_class_hits', 'top_chemical_class',\n",
    "                                             'top_chemical_class_proportion', 'top_chemical_subclass',\n",
    "                                             'top_chemical_subclass_proportion', 'taxonomic_hits', \n",
    "                                             'taxonomic_level', 'H-index', 'richness', 'top_taxa', \n",
    "                                             'top_taxa_proportion']):\n",
    "    for bgc in G.nodes:\n",
    "        if bgc in df.index:\n",
    "            G.nodes[bgc]['node_type'] = \"BiG-FAM GCFs\"\n",
    "            G.nodes[bgc]['text'] = f\"GCF {bgc}<br>size: {df.loc[bgc, 'bgc_member']}<br>top_chemical_class: {df.loc[bgc, 'top_chemical_class']} ({df.loc[bgc, 'top_chemical_class_proportion']:.0%})\\\n",
    "<br>top_chemical_subclass: {df.loc[bgc, 'top_chemical_subclass']} ({df.loc[bgc, 'top_chemical_subclass_proportion']:.0%})\\\n",
    "<br>top_taxa: {df.loc[bgc, 'top_taxa']} ({df.loc[bgc, 'top_taxa_proportion']:.0%})\"\n",
    "            for c in columns:\n",
    "                G.nodes[bgc][c] = df.loc[bgc, c]\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd034c15-182e-4d65-8266-3392a2a0ec60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_bigfam_antismash(G, antismash_region_path, columns = [\"genome_id\", \"product\", \"contig_edge\", \"region_length\", \"most_similar_known_cluster_id\", \"most_similar_known_cluster_description\", \"most_similar_known_cluster_type\", \"similarity\"]):\n",
    "    df_antismash = pd.read_csv(antismash_region_path, index_col=0)\n",
    "    for bgc in G.nodes:\n",
    "        if bgc in df_antismash.index:\n",
    "            G.nodes[bgc]['node_type'] = \"BGC\"\n",
    "            G.nodes[bgc]['text'] = f\"{bgc}<br>{df_antismash.loc[bgc, 'product']}\"\n",
    "            if 'most_similar_known_cluster_description' in df_antismash.columns:\n",
    "                G.nodes[bgc]['text'] = G.nodes[bgc]['text'] + f\"<br>{df_antismash.loc[bgc, 'most_similar_known_cluster_description']}\"\n",
    "            for c in df_antismash.columns:\n",
    "                if c in columns:\n",
    "                    G.nodes[bgc][c] = df_antismash.loc[bgc, c]\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8923e90-4526-4146-ba33-6cff84cf2f7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_edge_trace(G, name):\n",
    "    edge_trace = go.Scatter(\n",
    "        x=[],\n",
    "        y=[],\n",
    "        name=name,\n",
    "        line=dict(width=0.5,color='#888'),\n",
    "        hoverinfo='none',\n",
    "        mode='lines')\n",
    "\n",
    "    for edge in G.edges():\n",
    "        x0, y0 = G.nodes[edge[0]]['pos']\n",
    "        x1, y1 = G.nodes[edge[1]]['pos']\n",
    "        edge_trace['x'] += tuple([x0, x1, None])\n",
    "        edge_trace['y'] += tuple([y0, y1, None])\n",
    "    return edge_trace\n",
    "\n",
    "def create_node_trace(G, node_trace_category, color, showtextlabel=False, nodesize=10, nodeopacity=0.8, nodesymbol=\"circle\", linewidth=1, linecolor=\"black\", textposition=\"top center\"):\n",
    "    if showtextlabel:\n",
    "        markermode = \"markers+text\"\n",
    "    else:\n",
    "        markermode = \"markers\"\n",
    "    node_trace = go.Scatter(\n",
    "            x=[],\n",
    "            y=[],\n",
    "            text=[],\n",
    "            textposition=textposition,\n",
    "            mode=markermode,\n",
    "            hoverinfo='text',\n",
    "            name=node_trace_category,\n",
    "            marker=dict(\n",
    "                symbol=nodesymbol,\n",
    "                opacity=nodeopacity,\n",
    "                showscale=False,\n",
    "                color=color,\n",
    "                size=nodesize,\n",
    "                line=dict(width=linewidth, color=linecolor)))\n",
    "\n",
    "    for node in G.nodes():\n",
    "        if G.nodes[node][\"node_trace\"] == node_trace_category:\n",
    "            x, y = G.nodes[node]['pos']\n",
    "            node_trace['x'] += tuple([x])\n",
    "            node_trace['y'] += tuple([y])\n",
    "            node_trace['text'] +=tuple([G.nodes[node]['text']])\n",
    "    return node_trace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53fb8b15-379c-4be1-827e-243602f89d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "report_dir = Path(\"../\")\n",
    "\n",
    "with open(report_dir / \"metadata/project_metadata.json\", \"r\") as f:\n",
    "    project_configuration = json.load(f)\n",
    "with open(report_dir / \"metadata/dependency_versions.json\", \"r\") as f:\n",
    "    dependency_version = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a457b21e-12a4-4465-bab1-0a9cf2eb309b",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_name = [i for i in project_configuration.keys()][0]\n",
    "antismash_version = dependency_version[\"antismash\"]\n",
    "\n",
    "query_dir = report_dir / f\"bigslice/query_as_{antismash_version}/\"\n",
    "df_bigfam_model = pd.read_csv(query_dir / \"gcf_summary.csv\")\n",
    "df_bigfam_hits = pd.read_csv(report_dir / f\"bigslice/query_as_{antismash_version}/query_network.csv\")\n",
    "antismash_region_path = report_dir / f\"tables/df_regions_antismash_{antismash_version}.csv\"\n",
    "df_annotated_bigfam_model = pd.read_csv(query_dir / \"gcf_annotation.csv\", index_col=0)\n",
    "df_network = pd.read_csv(query_dir / \"query_network.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d2b11e1-1cd9-4443-bc77-6d4bbb7da529",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filtering\n",
    "#df_network = df_network[df_network[\"rank\"] < 3] # get first degree only\n",
    "df_annotated_bigfam_model = df_annotated_bigfam_model[df_annotated_bigfam_model[\"bgc_member\"] < 15000] # remove model with really large size\n",
    "df_network = df_network[df_network[\"gcf_id\"].isin(df_annotated_bigfam_model.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a7732c-6a11-43a9-b2bd-4406db0be290",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cut = df_annotated_bigfam_model[df_annotated_bigfam_model.top_taxa_proportion <= 0.3]\n",
    "\n",
    "G = generate_bigfam_network(df_network)\n",
    "G = annotate_bigfam_antismash(G, antismash_region_path)\n",
    "G = annotate_bigfam_models(G, df_annotated_bigfam_model)\n",
    "G_raw = G.copy()\n",
    "\n",
    "# position nodes\n",
    "pos = nx.nx_agraph.graphviz_layout(G)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ad9896-2043-49b8-9e39-aab12f57f5fa",
   "metadata": {},
   "source": [
    "## Summary Result\n",
    "### Distribution of the assigned genus from BiG-FAM Model hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26c9b54f-c1ed-4180-ae32-f9635f958509",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the number of GCFs with representation across different genera\n",
    "df_genus_dist = pd.DataFrame(index=df_annotated_bigfam_model.index)\n",
    "for gcf_id in df_annotated_bigfam_model.index:\n",
    "    taxa = str(df_annotated_bigfam_model.loc[gcf_id, \"taxa_distribution\"])\n",
    "    taxa_dict = ast.literal_eval(taxa)\n",
    "    for genus in taxa_dict.keys():\n",
    "        df_genus_dist.loc[gcf_id, genus] = taxa_dict[genus]\n",
    "df_genus_dist.fillna(0, inplace=True)\n",
    "\n",
    "# Genera represented by most number of GCFs\n",
    "df_genus_dist_binary = df_genus_dist > 0 \n",
    "df_genus_dist_binary = df_genus_dist_binary * 1\n",
    "\n",
    "# Assuming df_genus_dist_binary.sum().sort_values(ascending=False)[:10] is stored in a variable named 'data'\n",
    "data = df_genus_dist_binary.sum().sort_values(ascending=False)[:10]\n",
    "\n",
    "# Convert Series to DataFrame for Altair plotting\n",
    "data_df = data.reset_index()\n",
    "data_df.columns = ['Genus', 'Counts']\n",
    "\n",
    "# Create Altair Chart\n",
    "chart = alt.Chart(data_df).mark_bar().encode(\n",
    "    y=alt.Y('Genus:N', sort='-x', title='Genera'),\n",
    "    x=alt.X('Counts:Q', title='Counts'),\n",
    "    color=alt.Color('Genus:N', legend=None)  # Color by Genera, but we don't need a legend here\n",
    ").properties(\n",
    "    title='Top 10 Genus Distribution',\n",
    "    width=700,\n",
    "    height=300\n",
    ")\n",
    "\n",
    "# Display the chart\n",
    "chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "087dfb47-9c8f-4c61-84d4-2d72b7ffcab4",
   "metadata": {},
   "source": [
    "### Diversity and Member Size of BiG-FAM model hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76e25859-6a93-4e5e-897d-f7fe60ba259d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bgc_hits = len([n for n in G_raw.nodes if G_raw.nodes[n]['node_type'] == 'BGC'])\n",
    "GCF_hits = len([n for n in G_raw.nodes if G_raw.nodes[n]['node_type'] != 'BGC'])\n",
    "shannon_non_zero = df_annotated_bigfam_model[df_annotated_bigfam_model['H-index'] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c50933f-688d-49ca-84c0-18ef7220815d",
   "metadata": {},
   "outputs": [],
   "source": [
    "shannon_min_non_zero_tax = ast.literal_eval(str(df_annotated_bigfam_model.loc[shannon_non_zero['H-index'].astype(float).idxmin(), \"taxa_distribution\"]))\n",
    "shannon_min_non_zero_tax = {k : v/sum(shannon_min_non_zero_tax.values()) for k,v in shannon_min_non_zero_tax.items()}\n",
    "shannon_min_non_zero_tax_clean = [f\"{k} ({v:.1%})\" for k,v in sorted(shannon_min_non_zero_tax.items(), key=lambda x:x[1], reverse=True)]\n",
    "\n",
    "def get_top_taxa(df, gcf):\n",
    "    result = f\"{df.loc[gcf, 'top_taxa']} ({df.loc[gcf, 'top_taxa_proportion'] * df.loc[gcf, 'bgc_member']:.0f} genomes)\"\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5251f9c-5dc9-4f61-a43d-d5e2558353c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = df_bigfam_hits.gcf_id.value_counts()\n",
    "for gcf in df_annotated_bigfam_model.index:\n",
    "    df_annotated_bigfam_model.loc[gcf, \"dataset_hits\"] = mapping[gcf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d0d36af-3a0c-4949-9f02-482e2595fa0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "domain = list(df_annotated_bigfam_model[df_annotated_bigfam_model[\"top_taxa_proportion\"] > 0.1][\"top_taxa\"].value_counts().to_dict().keys())\n",
    "domain.append(\"Other\")\n",
    "\n",
    "r = []\n",
    "range_ = [\"#264653\", \"#287271\", \"#2a9d8f\", \"#8ab17d\", \"#e9c46a\", \"#f4a261\", \"#ee8959\", \"#e76f51\", \"#edede9\"]\n",
    "for num, d in enumerate(domain):\n",
    "    if num < len(range_):\n",
    "        r.append(range_[num])\n",
    "    else:\n",
    "        r.append(\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5882dd62-7c22-432d-ac99-6fe2203c3217",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = df_annotated_bigfam_model.copy()\n",
    "source = source.reset_index().rename(columns={\"gcf_id\":\"BiG-FAM_id\"})\n",
    "for i in source.index:\n",
    "    if source.loc[i, \"top_taxa_proportion\"] <= 0.5:\n",
    "        source.loc[i, \"top_taxa\"] = \"Other\"\n",
    "        \n",
    "chart_one = alt.Chart(source).mark_point().encode(\n",
    "    alt.Y('H-index:Q',\n",
    "          scale=alt.Scale(domain=(-0.5, 5)),\n",
    "          axis=alt.Axis(title=\"Shannon Index (H')\")\n",
    "         ),\n",
    "    alt.X('bgc_member:Q',\n",
    "          scale=alt.Scale(type=\"log\"),\n",
    "          axis=alt.Axis(title=\"Member Size\")\n",
    "         ),\n",
    "    alt.Size('dataset_hits',\n",
    "             scale=alt.Scale(type='pow'),# domain=(1, 30)), \n",
    "             title=\"Number of hits in dataset\"\n",
    "            ),\n",
    "    alt.Color(\"top_taxa:N\", scale=alt.Scale(domain=domain, range=r), title=\"Top Genus (>=50%)\"),\n",
    "    tooltip=['BiG-FAM_id', 'bgc_member', 'chemical_class_hits', 'top_chemical_class', alt.Tooltip('top_chemical_class_proportion', format='.1%'), \n",
    "             'top_chemical_subclass', alt.Tooltip('top_chemical_subclass_proportion', format='.1%'),\n",
    "             'taxonomic_level', 'richness', 'top_taxa', alt.Tooltip('top_taxa_proportion', format='.1%'), \n",
    "             alt.Tooltip('H-index:Q', format='.2f')],\n",
    ").mark_point(\n",
    "    filled=True,\n",
    "    stroke='black',\n",
    "    strokeWidth=0.5,\n",
    "    opacity=0.8,\n",
    "    size=1000\n",
    ").configure_header(\n",
    "    title=None,\n",
    "    labels=False\n",
    ").configure_axis(\n",
    "    labelFontSize=10,\n",
    "    titleFontSize=12\n",
    ").configure_legend(\n",
    "    labelFontSize=10,\n",
    "    titleFontSize=12,\n",
    ").configure_view(\n",
    "    continuousHeight=500,\n",
    "    continuousWidth=500,\n",
    ").configure_legend(\n",
    "  orient='right'\n",
    ")\n",
    "\n",
    "chart_one"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb27456a-066f-476a-83d1-08bd307f28f8",
   "metadata": {},
   "source": [
    "### Network of BiG-FAM model hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a760c64-c7e2-442c-92e3-ef1f1eccda63",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_category = 'node_type'\n",
    "colormap = {\"BiG-FAM GCFs\" : \"red\",\n",
    "           \"BGC\" : \"blue\"}\n",
    "node_trace_category = {}\n",
    "\n",
    "for node in G.nodes:\n",
    "    nodeshape = \"circle\"\n",
    "    if node in df_annotated_bigfam_model.index:\n",
    "        G.nodes[node]['node_type'] = \"BiG-FAM GCFs\"\n",
    "    if G.nodes[node]['node_type'] == \"BiG-FAM GCFs\":\n",
    "        nodeshape = \"diamond\"\n",
    "        color = colormap[\"BiG-FAM GCFs\"]\n",
    "        cat = \"BiG-FAM GCFs\"\n",
    "    else:            \n",
    "        cat = G.nodes[node][color_category]\n",
    "        color = colormap[cat]\n",
    "        if 'similarity' in G.nodes[node].keys():\n",
    "            if G.nodes[node]['similarity'] > 0.8:\n",
    "                cat = cat + \" \" + \"Known (antiSMASH) > 80% similarity\"\n",
    "                nodeshape = \"circle\"\n",
    "            elif G.nodes[node]['similarity'] > 0.4:\n",
    "                cat = cat + \" \" + \"Known (antiSMASH) > 40% similarity\"\n",
    "                nodeshape = \"triangle-up\"\n",
    "            elif G.nodes[node]['similarity'] < 0.4:\n",
    "                cat = cat + \" \" + \"Unknown (antiSMASH) < 40% similarity\"\n",
    "                nodeshape = \"triangle-down\"\n",
    "        else:\n",
    "            cat = cat + \" \" + \"Unknown\"\n",
    "            nodeshape = \"circle\"\n",
    "        \n",
    "        linecolor = \"black\"\n",
    "    \n",
    "    G.nodes[node]['node_trace'] = cat\n",
    "    node_trace_category[cat] = {\"nodecolor\" : color,\n",
    "                                \"nodeshape\" : nodeshape,\n",
    "                                \"linecolor\" : linecolor}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb871352-8602-438d-8404-82e36f2c1afd",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_trace = create_edge_trace(G, \"bigslice_similarity\")\n",
    "traces = [edge_trace]\n",
    "for cat in node_trace_category.keys():\n",
    "    nodeopacity = 0.8\n",
    "    color = node_trace_category[cat][\"nodecolor\"]\n",
    "    nodeshape = node_trace_category[cat][\"nodeshape\"]\n",
    "    linecolor = node_trace_category[cat][\"linecolor\"]\n",
    "    node_trace = create_node_trace(G, cat, color, nodesymbol=nodeshape, nodeopacity=nodeopacity, nodesize=10, linewidth=1, linecolor=linecolor)\n",
    "    traces.append(node_trace)\n",
    "    \n",
    "fig2 = go.Figure(data=traces,\n",
    "                layout=go.Layout(\n",
    "                    paper_bgcolor='rgba(0,0,0,0)',\n",
    "                    plot_bgcolor='rgba(0,0,0,0)',\n",
    "                    showlegend=True,\n",
    "                    hovermode='closest',\n",
    "                    margin=dict(b=20,l=5,r=5,t=40),\n",
    "                    xaxis=dict(showgrid=False, zeroline=False, showticklabels=False),\n",
    "                    yaxis=dict(showgrid=False, zeroline=False, showticklabels=False),\n",
    "                    width=1200, height=800))\n",
    "\n",
    "outfile = Path(\"assets/figures/bigfam-query.html\")\n",
    "outfile.parent.mkdir(parents=True, exist_ok=True)\n",
    "fig2.write_html(outfile)\n",
    "\n",
    "display(HTML(filename=str(outfile)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8fe1ca-30f9-472a-9e01-4e9db61825b8",
   "metadata": {},
   "source": [
    "## References\n",
    "<font size=\"2\">\n",
    "{% for i in project().rule_used['query-bigslice']['references'] %}\n",
    "  - *{{ i }}*\n",
    "{% endfor %}\n",
    "</font>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
